File: frontend/src/constants/permissions.js
// Exact mirror of backend/permissions.py — RBAC single source of truth (frontend).

export const MODULE_ACCESS = {
  "dashboard":       new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "marketplace":     new Set(["Founder"]),
  "task-board":      new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "opportunity-hub": new Set(["Founder", "Admin", "Manager", "Employee"]),
  "employees":       new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "wavygo-connect":  new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "company-vault":   new Set(["Founder"]),
  "finance":         new Set(["Founder"]),
  "crm":             new Set(["Founder", "Admin", "Manager"]),
  "marketing":       new Set(["Founder", "Admin", "Manager"]),
  "analytics":       new Set(["Founder", "Admin", "Manager"]),
  "calendar":        new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "wavygo-ai":       new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "about-wavygo":    new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "activity-logs":   new Set(["Founder", "Admin", "Manager"]),
  "notifications":   new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "settings":        new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
};

export const SIDEBAR_LABELS = {
  "task-board":      { Employee: "My Tasks", Intern: "Assigned Tasks" },
  "opportunity-hub": { Employee: "My Opportunities" },
  "employees":       { Employee: "My Workspace", Intern: "My Workspace" },
  "settings":        { Employee: "My Profile", Intern: "My Profile" },
};

export const ACTIONS = {
  "auth.register":               new Set(["Founder", "Admin"]),
  "auth.change_own_password":    new Set(["Founder", "Admin", "Manager"]),
  "auth.reset_other_password":   new Set(["Founder", "Admin"]),

  "user.invite.founder":         new Set([]),
  "user.invite.admin":           new Set(["Founder"]),
  "user.invite.manager":         new Set(["Founder", "Admin"]),
  "user.invite.employee":        new Set(["Founder", "Admin"]),
  "user.invite.intern":          new Set(["Founder", "Admin"]),
  "user.delete":                 new Set(["Founder"]),
  "user.edit_others":            new Set(["Founder", "Admin", "Manager"]),
  "user.edit_self":              new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),

  "marketplace.any":             new Set(["Founder"]),

  "task.create":                 new Set(["Founder", "Admin", "Manager", "Employee"]),
  "task.edit_any":               new Set(["Founder", "Admin", "Manager"]),
  "task.edit_own":               new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "task.delete":                 new Set(["Founder", "Admin"]),
  "task.assign":                 new Set(["Founder", "Admin", "Manager"]),
  "task.comment":                new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),

  "opportunity.create":          new Set(["Founder", "Admin", "Manager"]),
  "opportunity.assign":          new Set(["Founder", "Admin", "Manager"]),
  "opportunity.edit_any":        new Set(["Founder", "Admin"]),
  "opportunity.edit_own":        new Set(["Founder", "Admin", "Manager", "Employee"]),
  "opportunity.delete":          new Set(["Founder"]),

  "employee.invite":             new Set(["Founder", "Admin"]),
  "employee.edit":               new Set(["Founder", "Admin", "Manager"]),
  "employee.view_directory":     new Set(["Founder", "Admin", "Manager"]),
  "employee.view_self":          new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "department.create":           new Set(["Founder", "Admin"]),

  "attendance.mark_self":        new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "attendance.mark_others":      new Set(["Founder", "Admin", "Manager"]),
  "leave.request_self":          new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
  "leave.approve":               new Set(["Founder", "Admin", "Manager"]),
  "performance.create":          new Set(["Founder", "Admin", "Manager"]),
  "performance.view_self":       new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),

  "connect.create_channel":      new Set(["Founder", "Admin", "Manager"]),
  "connect.create_announcement": new Set(["Founder", "Admin"]),
  "connect.post_announcement":   new Set(["Founder", "Admin"]),
  "connect.send_dm":             new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),

  "activity.view_all":           new Set(["Founder", "Admin"]),
  "activity.view_team":          new Set(["Manager"]),

  "settings.company_edit":       new Set(["Founder"]),
  "settings.roles_view":         new Set(["Founder", "Admin", "Manager", "Employee", "Intern"]),
};

export function canViewModule(role, key) {
  const set = MODULE_ACCESS[key];
  return !!set && set.has(role);
}

export function can(role, action) {
  const set = ACTIONS[action];
  if (!set) throw new Error(`Unknown action: ${action}`);
  return set.has(role);
}

export function sidebarLabel(moduleKey, defaultLabel, role) {
  return (SIDEBAR_LABELS[moduleKey] && SIDEBAR_LABELS[moduleKey][role]) || defaultLabel;
}

File: backend/permissions.py
"""Central RBAC permission matrix for WavyGo OS.

Single source of truth for module visibility and per-action permissions.
Frontend mirrors this in src/constants/permissions.js.
"""

MODULE_ACCESS = {
    "dashboard":       {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "marketplace":     {"Founder"},
    "task-board":      {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "opportunity-hub": {"Founder", "Admin", "Manager", "Employee"},
    "employees":       {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "wavygo-connect":  {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "company-vault":   {"Founder"},
    "finance":         {"Founder"},
    "crm":             {"Founder", "Admin", "Manager"},
    "marketing":       {"Founder", "Admin", "Manager"},
    "analytics":       {"Founder", "Admin", "Manager"},
    "calendar":        {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "wavygo-ai":       {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "about-wavygo":    {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "activity-logs":   {"Founder", "Admin", "Manager"},
    "notifications":   {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "settings":        {"Founder", "Admin", "Manager", "Employee", "Intern"},
}

SIDEBAR_LABELS = {
    "task-board":      {"Employee": "My Tasks", "Intern": "Assigned Tasks"},
    "opportunity-hub": {"Employee": "My Opportunities"},
    "employees":       {"Employee": "My Workspace", "Intern": "My Workspace"},
    "settings":        {"Employee": "My Profile", "Intern": "My Profile"},
}

ACTIONS = {
    "auth.register":                {"Founder", "Admin"},
    "auth.change_own_password":     {"Founder", "Admin", "Manager"},
    "auth.reset_other_password":    {"Founder", "Admin"},

    "user.invite.founder":          set(),
    "user.invite.admin":            {"Founder"},
    "user.invite.manager":          {"Founder", "Admin"},
    "user.invite.employee":         {"Founder", "Admin"},
    "user.invite.intern":           {"Founder", "Admin"},
    "user.delete":                  {"Founder"},
    "user.edit_others":             {"Founder", "Admin", "Manager"},
    "user.edit_self":               {"Founder", "Admin", "Manager", "Employee", "Intern"},

    "marketplace.any":              {"Founder"},

    "task.create":                  {"Founder", "Admin", "Manager", "Employee"},
    "task.edit_any":                {"Founder", "Admin", "Manager"},
    "task.edit_own":                {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "task.delete":                  {"Founder", "Admin"},
    "task.assign":                  {"Founder", "Admin", "Manager"},
    "task.comment":                 {"Founder", "Admin", "Manager", "Employee", "Intern"},

    "opportunity.create":           {"Founder", "Admin", "Manager"},
    "opportunity.assign":           {"Founder", "Admin", "Manager"},
    "opportunity.edit_any":         {"Founder", "Admin"},
    "opportunity.edit_own":         {"Founder", "Admin", "Manager", "Employee"},
    "opportunity.delete":           {"Founder"},

    "employee.invite":              {"Founder", "Admin"},
    "employee.edit":                {"Founder", "Admin", "Manager"},
    "employee.view_directory":      {"Founder", "Admin", "Manager"},
    "employee.view_self":           {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "department.create":            {"Founder", "Admin"},

    "attendance.mark_self":         {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "attendance.mark_others":       {"Founder", "Admin", "Manager"},
    "leave.request_self":           {"Founder", "Admin", "Manager", "Employee", "Intern"},
    "leave.approve":                {"Founder", "Admin", "Manager"},
    "performance.create":           {"Founder", "Admin", "Manager"},
    "performance.view_self":        {"Founder", "Admin", "Manager", "Employee", "Intern"},

    "connect.create_channel":       {"Founder", "Admin", "Manager"},
    "connect.create_announcement":  {"Founder", "Admin"},
    "connect.post_announcement":    {"Founder", "Admin"},
    "connect.send_dm":              {"Founder", "Admin", "Manager", "Employee", "Intern"},

    "activity.view_all":            {"Founder", "Admin"},
    "activity.view_team":           {"Manager"},

    "settings.company_edit":        {"Founder"},
    "settings.roles_view":          {"Founder", "Admin", "Manager", "Employee", "Intern"},
}


def can_view_module(role, key):
    return role in MODULE_ACCESS.get(key, set())


def can(role, action):
    if action not in ACTIONS:
        raise ValueError(f"Unknown action: {action}")
    return role in ACTIONS[action]


def sidebar_label(module_key, default_label, role):
    return SIDEBAR_LABELS.get(module_key, {}).get(role, default_label)
